README.md

```diff
@@ -441,6 +441,30 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.container-log-max-files`: The maximum number of container log files that can be present for a container.
 * `settings.kubernetes.cpu-manager-policy`: Specifies the CPU manager policy. Possible values are `static` and `none`. Defaults to `none`. If you want to allow pods with certain resource characteristics to be granted increased CPU affinity and exclusivity on the node, you can set this setting to `static`. You should reboot if you change this setting after startup - try `apiclient reboot`.
 * `settings.kubernetes.cpu-manager-reconcile-period`: Specifies the CPU manager reconcile period, which controls how often updated CPU assignments are written to cgroupfs. The value is a duration like `30s` for 30 seconds or `1h5m` for 1 hour and 5 minutes.
+* `settings.kubernetes.credential-providers`: Contains a collection of Kubelet image credential provider settings.
+
+  Example user data for setting up image credential providers:
+
+  ```toml
+  [settings.kubernetes.credential-providers]
+  # AWS config settings for the "default" profile
+  # Example command: "base64 ~/.aws/config"
+  aws-config = "base64_encoded_content"
+  # Example command: "base64 ~/.aws/credentials"
+  # (optional - credentials can be included in aws-config content)
+  aws-credentials = "base64_encoded_content"
```

Comment:
My only real misgiving with this change is with these fields - it's going to be rather awkward if we end up needing AWS creds to be specified in some other context ("just put your creds for CloudWatch logging in the kubelet credential providers field").

What do you think about this as an alternative?
* Add these fields to `AwsSettings`
* Add `AwsSettings` to all models
* Conditionalize the `region` field in the struct with `#[cfg(variant_platform = "aws")]`

This would also let us simplify `[[settings.kubernetes.credential-providers.providers]]` down to just `[[settings.kubernetes.credential-providers]]` which would be more intuitive.

The main drawback I see is that we'd need to keep in mind whether a struct field in `AwsSettings` was "inside AWS" or "anywhere" in the future, but that seems OK.

Reply:
That was my original take, but a certain tech lead asked for these settings to be a base64 encoded blob. ;)

I was/am concerned about any potential future functionality that would require AWS creds. If we wanted to keep with the blob route, I see a couple options:

- Change location to something specific for credential providers, then pass along environment variables to get it to read the credentials from somewhere other than `~/.aws`
- Allow multiple sub-settings to use AWS creds, but merge them into one file and require each be given a different profile name to use

The challenge with using explicit structs for the AWS creds is we will end up having three different options (at least so far) for how these AWS creds are configured:

- Left blank so instance role is used
- Populated with normal content a la `aws configure`
- Configured to use IAM Roles Anywhere with content like the following:
  ```
   [default]
   credential_process = ./aws_signing_helper credential-process \
      --certificate /path/to/certificate \
      --private-key /path/to/private-key \
      --trust-anchor-arn arn:aws:rolesanywhere:region:account:trust-anchor/TA_ID \
      --profile-arn arn:aws:rolesanywhere:region:account:profile/PROFILE_ID \
      --role-arn arn:aws:iam::account:role/role-name-with-path
  ```

We could expand our `settings.aws` to accomodate either of those (`settings.aws.config` and `settings.aws.credential-process`?). Maybe also require a profile name to be assigned to them, then require the credential provider to also provide the profile name to help differentiate which of the configured settings to use.

A few caveats here, so let me know if we should schedule a call to talk them through.

Reply:
Adding what was discussed here to capture it for everyone.

We decided a good way to go would be to add an `enum` value that could either be a single `aws.profile` string, or a lookup of `aws.profile.[name]` values. So something like:

```toml
[settings.aws]
profile = default
```

or

```toml
[settings.aws.profile]
ecr-credential-provider = foo
future-service = default
```

I just pushed up changes to support that, though I'm running in to some issues. If I set the latter, then try to go back and set a single value, `thar-be-settings` complains that `profile` is a directory. It looks like it would need to have the smarts to delete this existing datastore location if someone wanted to switch back and forth between the two.

I'm also having problems pulling in `settings.aws.profile` into the `credential-provider-config.yaml` file. For some reason even though I see it set with `apiclient get settings.aws`, the evaluation always fails in the template so it never sets its `AWS_PROFILE` setting.

One more thing I noticed. With moving the `config` and `credentials` under `settings.aws`, for some reason it isn't detecting when those settings change and writing out the files under `/root/.aws`. I have those files still under kubernetes-settings, for lack of a more obvious correct place, so I would have expected changing those values to write out the rendered template and restart the kubelet service. It's only after changing `settings.kubernetes.credential-provider` settings that they get written out and the service gets restarted.

So still several things wrong with the current changes here. It feels like things are close, so I'm hoping I'm just missing some of the glue to pull these parts together.

Reply:
I was having a hard time getting the switch between the single `profile` as a string and `profile` as a mapping of names to values. I think it might take some changes to schnauzer or thar-be-settings to allow it since it was an error that `profile` was a directory when it expected it to be a file. So for now I have set `profile` to only allow a single profile name to use for all services. We should be able to follow up with the combined string or key/value approach if and when we determine we actually need that.

With that change, I think this is very close. I still have the issue that any updates to `settings.aws` don't appear to trigger a restart of kubelet nor the writing out of the `aws-config` or `aws-credentials` templates under `/root/.aws`. I'm sure I'm just missing some wiring in the right place, but I may need some help figuring out where that would be.

Reply:
> Conditionalize the region field in the struct with #[cfg(variant_platform = "aws")]

Hmmmm... not sure we should take our conditional compilation to that level of granularity.

Reply:
I did try it out and it caused some compilation errors. I am not sure if there are any non-AWS local use cases where region might be needed to, so it may be safer to keep that for now.